src/client/src/components/chatpage.js

```diff
@@ -0,0 +1,128 @@
+const React = require('react')
+const {BrowserRouter, Link} = require('react-router-dom')
+const ConnectStatus = require('./connectStatus')
+const Room = require('./room')
+const socket = require('../utils/sockets/ws-routing')
+const protocols = require('../utils/io-protocol')
+
+/*
+Consulted documentation:
+https://github.com/facebook/create-react-app/issues/2260
+https://github.com/socketio/socket.io/issues/1942#issuecomment-299764672
+https://github.com/socketio/socket.io/issues/2294
+
+https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/Redirect.md
+https://socket.io/docs/client-api/
+
+*/
+
+function updateState(key, array, newItems){
```

Comment:
Agreed, I was considering some abstraction to clean up the array concatenation since it was getting rather long, but then messages data structure changed so that it no longer used arrays explicitly, thus nullifying the need for this abstraction. I'll remove it.